indykite/data_source_application.go

```diff
@@ -72,46 +74,94 @@ func dataSourceApplicationList() *schema.Resource {
 	}
 }
 
+// lookupApplicationByName finds an application by name within an app space.
+func lookupApplicationByName(
+	ctx context.Context,
+	clientCtx *ClientContext,
+	data *schema.ResourceData,
+	name string,
+) (ApplicationResponse, diag.Diagnostic) {
+	// Get project ID from either project_id or app_space_id (backward compatibility)
+	projID := data.Get(projectIDKey).(string)
+	if projID == "" {
+		projID = data.Get(appSpaceIDKey).(string)
+	}
+
+	// Validate that at least one of app_space_id or project_id is provided when using name
+	if projID == "" {
+		return ApplicationResponse{}, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "\"name\": all of `app_space_id, name` must be specified, or use `project_id, name` instead",
+		}
+	}
+
+	var listResp ListApplicationsResponse
+	err := clientCtx.GetClient().Get(ctx, "/applications?project_id="+projID+"&search=", &listResp)
+	if err != nil {
+		return ApplicationResponse{}, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  fmt.Sprintf("failed to list applications: %v", err),
+		}
+	}
+
+	// Find application by name
+	for i := range listResp.Applications {
+		if listResp.Applications[i].Name == name {
+			return listResp.Applications[i], diag.Diagnostic{}
+		}
+	}
+
+	return ApplicationResponse{}, diag.Diagnostic{
+		Severity: diag.Error,
+		Summary: fmt.Sprintf(
+			"application with name '%s' not found in app space '%s'",
+			name, projID),
+	}
+}
+
 func dataApplicationReadContext(ctx context.Context, data *schema.ResourceData, meta any) diag.Diagnostics {
 	var d diag.Diagnostics
 	clientCtx := getClientContext(&d, meta)
 	if d.HasError() {
 		return d
 	}
 
-	req := &configpb.ReadApplicationRequest{}
-	if name, exists := data.GetOk(nameKey); exists {
-		req.Identifier = &configpb.ReadApplicationRequest_Name{
-			Name: &configpb.UniqueNameIdentifier{
-				Name:     name.(string),
-				Location: data.Get(appSpaceIDKey).(string),
-			},
-		}
-	} else if id, ok := data.GetOk(applicationIDKey); ok {
-		req.Identifier = &configpb.ReadApplicationRequest_Id{
-			Id: id.(string),
+	ctx, cancel := context.WithTimeout(ctx, data.Timeout(schema.TimeoutRead))
+	defer cancel()
```

Comment:
TODO: you don't need to handle timeouts yourself. Terraform handles that for you.